crates/pathfinder/src/rpc/api.rs

```diff
@@ -104,16 +105,115 @@ impl RpcApi {
+        }
+
+        if let BlockHashOrTag::Tag(Tag::Pending) = block_hash {
+            // Pending request is always forwarded
+            return Ok(self
+                .sequencer
+                .storage(contract_address, key, block_hash)
+                .await?);
+        }
+
+        // We cannot just return Error::Internal (-32003) in cases which are not covered by starknet RPC API spec
+        // as jsonrpsee reserved it for internal subscription related errors only, so we resort to
+        // CallError::Custom with the same code value and message as Error::Internal. This way we can still provide
+        // an "Internal server error" but with additional context.
+        //
+        // This error is used for all instances of operations that are not explicitly specified in the StarkNet spec.
+        // See https://github.com/starkware-libs/starknet-specs/blob/master/api/starknet_api_openrpc.json
+        let internal_server_error = |e: anyhow::Error| {
+            Error::Call(CallError::Custom {
+                code: jsonrpsee::types::v2::error::INTERNAL_ERROR_CODE,
+                message: format!("{}: {}", jsonrpsee::types::v2::error::INTERNAL_ERROR_MSG, e),
+                data: None,
+            })
+        };
+
+        let mut db = self
+            .storage
+            .connection()
+            .context("Opening database connection")
+            .map_err(internal_server_error)?;
+        let tx = db
+            .transaction()
+            .context("Creating database transaction")
+            .map_err(internal_server_error)?;
+
+        // Use internal_server_error to indicate that the process of querying for a particular block failed,
+        // which is not the same as being sure that the block is not in the db.
+        let global_root = match block_hash {
+            BlockHashOrTag::Hash(hash) => GlobalStateTable::get_root_at_block_hash(&tx, hash)
+                .map_err(internal_server_error)?,
+            BlockHashOrTag::Tag(tag) => match tag {
+                Tag::Latest => {
+                    GlobalStateTable::get_latest_root(&tx).map_err(internal_server_error)?
+                }
+                Tag::Pending => unreachable!("Pending has already been handled above"),
+            },
+        }
+        // Since the db query succeeded in execution, we can now report if the block hash was indeed not found
+        // by using a dedicated error code from the RPC API spec
+        .ok_or_else(|| Error::from(ErrorCode::InvalidBlockHash))?;
+
+        let global_state_tree = GlobalStateTree::load(&tx, global_root)
+            .context("Global state tree")
+            .map_err(internal_server_error)?;
+
+        // There is a dedicated error code for a non-existent contract in the RPC API spec, so use it.
+        let contract_state_hash = global_state_tree
+            .get(contract_address)
+            .map_err(|_| Error::from(ErrorCode::ContractNotFound))?;
```

Comment:
This is the way to check for non-existent key:
```suggestion
        let contract_state_hash = global_state_tree
            .get(contract_address)
            .context("Get contract state hash from global state tree")
            .map_err(internal_server_error)?;
            
        if contract_state_hash == StarkHash::ZERO {
            return Err(Error::from(ErrorCode::ContractNotFound));
        }
```

Reply:
Ahh yes, you're right :mechanical_arm: